# Le Wagon test technique TAs Data Science

Vous disposez de 2h pour compléter ce notebook en expliquant votre démarche et vos observations.
Gérez votre temps comme il vous semble le plus pertinent, la clarté des explication importe autant que la complétion totale du notebook.

**Format de rendu :**
- Lien vers un repository Github public.
- Ce lien devra être envoyé à Laura Mantovi et Agathe Sarran à 12h00.
- Le repository devra contenir votre notebook renommé comme suit : prenom_nom_test_TA_DS.ipynb
- Vous pourrez supprimer ce repository ou le passer en privé à partir de 17h.

Ce notebook doit être éxécutable dans l'environnement virtuel **lewagon** utilisé lors de votre bootcamp.

In [3]:
# ICI seront présent tout les imports utilisées
import pandas as pd

### Les données

Le fichier **data_weather.csv** comprend des données issues de plusieurs stations météo. Voici un rapide descriptif des colonnes:
- date: date d'enregistrement
- numer_sta: numéro unique de la station
- Latitude: Latitude
- Longitude: Longitude
- Altitude: Altitude
- pmer: pression au niveau de la mer (hpa)
- dd: direction du vent a 10m (sur 360°)
- ff: force du vent a 10m (m/s)
- t: température sous abri (°C)
- u: humidité
- ssfrai: unknown
- rr3: précipitations
- pres: pression à l'altitude de la station  (hpa)
- dd_sin: composante sin de direction du vent
- dd_cos: composante cos de direction du vent

In [5]:
df = pd.read_csv("data/data_weather.csv",
                 header=0,
                 names=["numer_sta", "latitude", "longitude", "altitude", "pmer", "dd", "t", "u", "ssfrai", "rr3", "pres", "dd_sin", "dd_cos"])

##### Data cleaning

Explorez rapidement les données et nettoyez les colonnes qui en ont besoin.<br>Documentez votre code pour expliquer vos choix.

In [ ]:
# Tout d'abord je regarde à quoi ressemble les données
df.head()

,,numer_sta,latitude,longitude,altitude,pmer,dd,t,u,ssfrai,rr3,pres,dd_sin,dd_cos
2010-01-01 00:00:00,7510.0,44.830667,-0.691333,47.0,99050.0,230.0,9.8,9.6,81.0,0.0,0,98410.0,-0.766044,-0.642788
2010-01-01 03:00:00,7510.0,44.830667,-0.691333,47.0,99160.0,250.0,11.8,8.7,87.0,0.0,0,98520.0,-0.939693,-0.342020
2010-01-01 06:00:00,7510.0,44.830667,-0.691333,47.0,99570.0,290.0,5.1,7.6,91.0,0.0,0,98920.0,-0.939693,0.342020
2010-01-01 09:00:00,7510.0,44.830667,-0.691333,47.0,99990.0,310.0,5.7,6.8,92.0,0.0,0,99340.0,-0.766044,0.642788
2010-01-01 12:00:00,7510.0,44.830667,-0.691333,47.0,100350.0,310.0,6.2,6.6,82.0,0.0,0,99690.0,-0.766044,0.642788


Ici, on peut déduire que notre data représente des données toutes les 3 heures de mesures météorologiques issues de différentes stations. 

In [ ]:
# On peut regarder les différentes
df.describe()

/Users/legregs/.pyenv/versions/3.12.9/envs/lewagon/lib/python3.12/site-packages/pandas/core/nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


,numer_sta,latitude,longitude,altitude,pmer,dd,t,u,ssfrai,rr3,pres,dd_sin,dd_cos
count,505857.000000,505857.000000,505857.000000,5.058570e+05,505857.000000,505857.000000,505857.000000,505857.000000,505857.000000,505857.000000,505791.000000,5.058570e+05,505857.000000
mean,45.478172,3.451461,193.196708,inf,185.492402,3.417473,13.266307,71.291632,0.000035,0.018035,99359.894818,-7.266871e-02,0.096686
std,1.502581,2.724426,217.479032,NaN,112.882202,2.350552,7.781763,18.898504,0.003194,0.133077,2597.553814,6.482709e-01,0.751743
min,43.437667,-1.608833,2.000000,1.059852e-205,0.000000,0.000000,-19.700000,1.000000,0.000000,0.000000,88160.000000,-1.000000e+00,-1.000000
25%,43.648833,1.175000,26.000000,1.012600e+05,80.000000,1.600000,7.700000,58.000000,0.000000,0.000000,98430.000000,-6.427876e-01,-0.642788
50%,45.726500,3.963167,151.000000,1.017500e+05,190.000000,3.000000,13.000000,74.000000,0.000000,0.000000,99960.000000,-2.449294e-16,0.173648
75%,47.150000,5.216000,263.000000,1.022900e+05,280.000000,4.600000,18.700000,87.000000,0.000000,0.000000,101120.000000,5.000000e-01,0.866025
max,47.614333,7.510000,871.000000,inf,360.000000,23.500000,42.000000,100.000000,0.990000,1.000000,104170.000000,1.000000e+00,1.000000


On voit ici déjà des valeurs aberrantes comme une altitude infinie, une moyenne de l'altitude infini mais aussi un count de pres inférieur à tout les autres.

In [ ]:
# Regardons les informations de ce DF, et on voit que pres contient des valeurs nulles.
df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 505857 entries, ('2010-01-01 00:00:00', 7510.0) to ('2021-07-31 21:00:00', 7650.0)
Data columns (total 13 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   numer_sta  505857 non-null  float64
 1   latitude   505857 non-null  float64
 2   longitude  505857 non-null  float64
 3   altitude   505857 non-null  float64
 4   pmer       505857 non-null  float64
 5   dd         505857 non-null  float64
 6   t          505857 non-null  float64
 7   u          505857 non-null  float64
 8   ssfrai     505857 non-null  float64
 9   rr3        505857 non-null  int64  
 10  pres       505791 non-null  float64
 11  dd_sin     505857 non-null  float64
 12  dd_cos     505857 non-null  float64
dtypes: float64(12), int64(1)
memory usage: 53.9+ MB


# PREPROCESS

* Doublons

Commençons par traiter les doublons ! A noter que l'on commence avec 505857 lignes

In [ ]:
# Pour savoir combien de doublons il y a : 89
print(df.duplicated().sum())

# On les retire !
df = df.drop_duplicates()
df.shape # (505768, 13), on a bien perdu 89 lignes

89


(505768, 13)

* Valeurs nulles

On peut maintenant se pencher sur les valeurs nulles

In [ ]:
# On regarde les valeurs nulles, et on retrouve bien les 65 vu précédemment dans pres
print(df.isnull().sum())

numer_sta     0
latitude      0
longitude     0
altitude      0
pmer          0
dd            0
t             0
u             0
ssfrai        0
rr3           0
pres         65
dd_sin        0
dd_cos        0
dtype: int64

In [31]:
# Regardons ces valeurs
df.loc[df['pres'].isna()]

,,numer_sta,latitude,longitude,altitude,pmer,dd,t,u,ssfrai,rr3,pres,dd_sin,dd_cos
2018-09-21 03:00:00,7434.0,45.861167,1.175000,402.0,101620.0,250.0,2.2,15.70,90.0,0.0,0,NaN,-9.396926e-01,-0.342020
2018-09-21 06:00:00,7434.0,45.861167,1.175000,402.0,101620.0,240.0,3.9,16.30,93.0,0.0,0,NaN,-8.660254e-01,-0.500000
2010-05-03 15:00:00,7643.0,43.577000,3.963167,2.0,100825.0,10.0,0.0,16.30,51.5,0.0,0,NaN,1.736482e-01,0.984808
2010-11-05 06:00:00,7643.0,43.577000,3.963167,2.0,102965.0,10.0,0.0,14.75,84.0,0.0,0,NaN,1.736482e-01,0.984808
2011-01-30 03:00:00,7643.0,43.577000,3.963167,2.0,101110.0,10.0,0.0,3.35,92.0,0.0,0,NaN,1.736482e-01,0.984808
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2014-03-05 15:00:00,7280.0,47.267833,5.088333,219.0,102110.0,10.0,5.7,11.80,43.0,0.0,0,NaN,1.736482e-01,0.984808
2021-04-28 12:00:00,7280.0,47.267833,5.088333,219.0,100420.0,200.0,1.8,10.50,71.5,0.0,0,NaN,-3.420201e-01,-0.939693
2013-03-17 03:00:00,7299.0,47.614333,7.510000,263.0,100655.0,260.0,0.0,3.85,61.5,0.0,0,NaN,-9.848078e-01,-0.173648
2020-05-18 21:00:00,7299.0,47.614333,7.510000,263.0,102110.0,260.0,0.0,17.95,54.5,0.0,0,NaN,-9.848078e-01,-0.173648


Ici plusieurs options s'offrent à nous: 

- Créer une nouvelle valeur pour ces valeurs nulles, c'est à dire les inpute.
Dans ce cas-ci, je prendrais probablement la moyenne de pres sur la journée.

- La deuxième solution est de supprimer ces lignes. Ces lignes ne réprésentent que 65 valeurs sur 505768, soit environ 0.0001 % des datas. De plus, en observant ces données manquantes ci-dessus, elles semblent rarement ce suivre (hormis les deux premiers).

Ici je vais donc choisir cette seconde option car la perte d'information est minime.

In [ ]:
# Enleve les na de notre jeu de données
df = df.dropna()
df.shape # Shape (505703, 13), on a bien perdu 65 lignes

(505703, 13)

* Outliers 

On peut ici s'intéresser aux outliers, et en particulier à la colonne altitude qu'on a déjà repéré

##### Data exploration

Quel est le numéro de la station qui a enregistré le plus de précipitations depuis 2010 ?<br>Dans quelle ville se situe-t-elle ?

Générez un fichier .csv contenant les moyennes des données météo mensuelles pour la station identifiée.

##### Data visualisation - *optionnel*

Pour chaque station d'enregistrement météo, représentez sur un graphique l'évolution de la moyenne mensuelle de la **température** et de l'**humidité** depuis 2010. <br>*La création des graphiques via l'utilisation d'objets **figure** et **axe** est un plus*.

### Approche Machine Learning

##### Modèles

Après avoir obtenu votre `baseline` explorez librement plusieurs algorithmes de Machine Learning pour prédire un événement pluvieux. Pensez à **expliquer votre méthodologie et les différentes étapes de votre raisonnement.**<br>
- sauvegardez vos scores dans un DataFrame `scores_tracking`
- sauvegardez votre meilleur modèle dans un fichier adapté
- enregistrez les paramètres de votre meilleur modèle dans une variable `best_model_params`

*Une attention particulière sera portée à votre méthode de préparation de données, au choix de(s) métrique(s) de performance, à la robustesse de vos scores et à la prévention de l'overfitting.*

Après avoir généré les valeurs nécessaires à l'affichage de vos **learning curves**, débugguez le code ci-dessous pour produire le graphique attendu.

In [ ]:
plt.plot(train_sizes, train_scores_mean, label = 'Training score')
plt.plot(test_sizes, test_scores_mean, 'Test score')
plt.ylabel('r2 score', fontsize = 14)
plt.xlabel('Test set size', fontsize = 14)
plt.title('Learning curves', fontsize = 18, y = 1.03)
plt.legend('center')

##### Etes-vous satifait de ces résultats? Comment les expliqueriez-vous ?


### Approche Deep Learning

Tentez d'améliorer vos prédictions grâce à un modèle de Deep Learning et documentez vos choix.